pravega/flink-connectors
# [issue-265] Reader side: Implement watermark with Pravega

**Change log description**

- Fix build error caused by breaking API change in `pravega-client`
- Implement reader with Pravega watermark. 
- Add watermark support in Stream Table Source of Table API. 
- Check only 1 stream linked with 1 `FlinkPravegaReader` when using `AssignerWithTimeWindow`.
- Add three test cases to make sure it works

**Purpose of the change**

- Support Pravega watermark in reader/source
- Merge into Flink `1.7` branch first because I start from it. Will cherry pick into `1.8` and `master` later

**What the code does**

- Implement Pravega watermark in both normal and Table API streaming reader/source
- We offer an `AssignerWithTimeWindow` interface for users. Users can use `withTimestampAndWatermark` to use Pravega watermark in the source function builder.
- We add watermark support `AssignerWithTimeWindow` class in the connectorProperties and integrate watermark in the underlying DataStreamSource of the Stream Table Source. Users can use `timestampFromSource` and `watermarkFromSource` in the `Rowtime` to use Pravega watermark.

**Limitations**

Due to the serialization validation for `connectorProperties`, only `public` `static-inner/outer` class implements `AssignerWithTimeWindow` is supported as a parameter of `withTimestampAndWatermark` in Table API stream table source. See: https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java#L489

**How to verify it**

Add three test cases:
1. TestHarness test for `FlinkPravegaReader`, watermark can periodically generated as expected.
2. Integration test for `FlinkPravegaReader`, guarantee no data loss and duplicate with windowing and watermark.
3. Interface validation test for Pravega `StreamTableSource` with watermark, validation passes.

`./gradlew clean build` pass.
